### PR TITLE
Support testing many variants of a block at runtime via dynamically generated templates

### PR DESCRIPTION
**Pull request by samueljseay**
### Changes proposed in this Pull Request:

This introduces a system in e2e tests for blocks that utilizes handlebars templates and wp-cli. It allows you to, at runtime, generate post content containing your block under test with many different attribute scenarios.

In this I reimplemented the e2e tests for stock status filter that I introduced in https://github.com/woocommerce/woocommerce/pull/44212

When considering all the things we could do, I didn't want to upfront generate lots of variations or anything like that. What we really want is a simple way to generate the block with dynamic attributes at runtime.  This is extremely flexible and with Handlebars being a flexible templating language, the things you can do are almost unlimited. In future though we could precompile templates or create some other pre-test system using the same concepts.

To get us started I added a simple helper that lets you render a wordpress block in a template and pass it attributes.

As you can see in my example template it looks like this:

```
{{#> wp-block blockName='woocommerce/product-filter-stock-status' attributes=attributes }}
{{/ wp-block }}
```

This also supports rendering child content in the block if you want to. 

This example only scratches the surface, you could use loops or other Handlebars features to generate virtually any template that suits you.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. See the CI pass
2. You could also try checking this out exploring locally, adding some templates of your own.

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

Add a system to support dynamic template testing in blocks e2e tests.

</details>

### CODE REVIEW

File: plugins/woo-ai/package.json [hunk 1]
```diff
@@ -125,6 +125,27 @@
 		"watch:build:project:wp-scripts": {
 			"command": "wp-scripts start",
 			"service": true
+		},
+		"dependencyOutputs": {
```
Comment (samueljseay) [inline plugins/woo-ai/package.json]: These are generated postinstall, so I wonder if it wasn't checked in for some reason.

File: plugins/woocommerce-blocks/tests/e2e/utils/cli.ts [hunk 2]
```diff
@@ -1,9 +1,17 @@
...
  * External dependencies
  */
-import { exec } from 'child_process';
+import { ExecException, exec } from 'child_process';
 
-export function cli( cmd, args = [] ) {
+export function cli(
+	cmd: string,
+	args = []
+): Promise< {
```
Comment (samueljseay) [inline plugins/woocommerce-blocks/tests/e2e/utils/cli.ts:9]: I ended up not using this due to needing to manipulate stdin, but this type improvement could be left in. TS couldn't infer the return type properly which was annoying.

File: plugins/woocommerce-blocks/tests/e2e/README.md [hunk 3]
```diff
@@ -88,7 +88,7 @@ pnpm run test:e2e:classic-theme
 pnpm run test:e2e:block-theme-with-templates
 ```
 
-_Note: All command parameters of `test:e2e` can be used for these commands too.
+\_Note: All command parameters of `test:e2e` can be used for these commands too.
```
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/tests/e2e/README.md:107]: typo here? ^^

File: plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+					id: postId,
+					url: `/${ title.toLowerCase().replace( /\s+/g, '-' ) }/`,
+					deletePost: () => deletePost( postId ),
+				} );
+			}
+		} );
+
+		if ( childProcess.stdin ) {
+			childProcess.stdin.write( templateContent );
+			childProcess.stdin.end();
```
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: Interesting!!!

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * Internal dependencies
+ */
+import {
+	GeneratedPost,
+	createPostFromTemplate,
+} from '../../utils/create-dynamic-content';
+
+const TEMPLATE_PATH = path.join( __dirname, 'stock-status.handlebars' );
```
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: Can we set the second param of `createPostFromTemplate` to detect relative path? like this:

```
		defaultBlockPost = await createPostFromTemplate(
			'Product Filter Stock Status Block',
			'./stock-status.handlebars',
			{}
		);
```

**Review by dinhtungdu — COMMENTED** [review]
Very promising! I left a minor comment but overall, this looks great to me!

File: plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	await cli(
+		`npm run wp-env run tests-cli -- wp post delete ${ id } --force`
+	);
+};
+
+// Creates a post as user 1.
+const createPost = async (
+	title: string,
+	templateContent: string
+): Promise< GeneratedPost > => {
```
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: Should we support more data related to the post being created? Like post type, category...

File: plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts (hunk 4 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: @dinhtungdu yeah I was forced into a corner using stdin here because we can't reference a file without copying it to the docker container and I couldn't get escaped strings to work.

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts (hunk 5 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts:15]: we certainly can 😄 happy to add that for convenience.

File: plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts (hunk 6 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: @dinhtungdu thats a good point, but I'm also not too concerned not to implement that, if someone else wants to extend this API they can do so easily when they need it?

I don't mind though, we certainly can pass an object representing more details about the post. I did wonder though, if we'd ever need it for this kind of testing. I guess if there were situations where our test was about how the block interacted with post meta and other attributes it might be relevant?

In theory even the title isn't really necessary here, but I've included it because it felt right to 😂

File: plugins/woocommerce-blocks/tests/e2e/README.md (hunk 3 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/README.md:107]: @dinhtungdu I thought that too when I first committed it, but its auto markdown formatting... 🤷🏻

Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/README.md:107]: Oh its because this isn't terminated with another `_` .. so its supposed to be the underscore character. I have no idea what the intent of this was originally.

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		await expect( stockStatuses.nth( 1 ) ).toHaveText( 'Out of stock' );
+	} );
+
+	test( 'Selecting a stock status filters the list of products', async ( {
+		page,
+	} ) => {
+		await page.goto( defaultBlockPost.url );
+
+		const outOfStockCheckbox = page.getByLabel( 'Out of stock' );
+		outOfStockCheckbox.click();
```
Comment (WunderBart) [inline plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]:
```suggestion
		await outOfStockCheckbox.click();
```

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+		// wait for navigation
+		await page.waitForURL( /.*filter_stock_status=outofstock.*/ );
+
+		const products = page.locator( '.wc-block-product' );
+		const productCount = await products.count();
+
+		expect( productCount ).toBe( 1 );
```
Comment (WunderBart) [inline plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: Let's use `toHaveCount` [to avoid flakiness](https://playwright.dev/docs/api/class-locator#locator-count):

```suggestion
		await expect( products ).toHaveCount( 1 );
```

File: plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts (hunk 6 above)
Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: I've only skimmed the framework here so far, but have we considered creating posts via a rest call? Like we do in gutenberg via RequestUtils fixture: https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-test-utils-playwright/src/request-utils/posts.ts#L60-L71

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+		await dropdownLocator.click();
+
+		await expect( page.getByText( 'In stock' ) ).toBeVisible();
+		await expect( page.getByText( 'Out of stock' ) ).toBeVisible();
+	} );
+
+	test.afterAll( async () => {
+		await defaultBlockPost.deletePost();
+		await dropdownBlockPost.deletePost();
+	} );
```
Comment (WunderBart) [inline plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: Will running multiple (independent?) tests against the same post be safe, stability-wise? Generally, if we'd like to run a multi-step scenario test against the same post, we can use the [`test.step` API](https://playwright.dev/docs/api/class-test#test-step). Otherwise, using a fresh post for each `test` would generally be recommended to avoid running the test in a (potentially) non-deterministic environment. To achieve that without duplicating code, we could consider turning `createPostFromTemplate` into a [Playwright Fixture](https://playwright.dev/docs/test-fixtures), where we'd also handle all the setup and teardown for each block post, saving precious LOCs. 😄 

Let me know if that makes sense!

**Review by WunderBart — COMMENTED** [review]
I've left a few comments/questions. Please bear with me since I'm very new to the `woocommerce` env! 🙏 😄

File: plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts (hunk 6 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: @WunderBart we certainly could but wouldn't it be faster via cli? We already have access to the CLI installed on the docker box itself.

I don't mind the suggestion of course, just wondering what would be the benefit here of using the rest API directly?

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts (hunk 7 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: Whoops! Thanks for spotting 🙇

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts (hunk 8 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: Great point thanks, I think we want to pass a locator to toHaveCount though right? I'll fix that. 🙇🏻

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts (hunk 9 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: Yes great question and point. It depends! And its the same with the existing templates we use (which are created once before the test framework is run). 

If you interacted with a block in the editor then certainly it would impact how it would behave on the front-end. To me it feels like its the responsibility of the developer to determine whether what they're doing should be torn down and when or be aware if what they're doing would have a side effect.

> Otherwise, using a fresh post for each test would generally be recommended to avoid running the test in a (potentially) non-deterministic environment. To achieve that without duplicating code, we could consider turning createPostFromTemplate into a [Playwright Fixture](https://playwright.dev/docs/test-fixtures), where we'd handle all the setup and teardown for each block post, saving precious LOCs. 😄

This API looks awesome thank you, I didn't realize playwright has so many useful APIs 🙇 

I think we should balance between using this fixture API and approaches like I've used here based on performance (and whether the post has side effects) mainly for performance reasons. Because creating the post and tearing it down takes some time to do every single test. But I'm still keen to be convinced I'm wrong here, tbh I haven't done a lot of PlayWright testing so maybe I'm wrong.

File: plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts (hunk 6 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: Actually I wonder if I'm making a bad assumption here. Maybe the rest API would perform better. certainly the code should be simpler.

Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: Good question. I don't know how it would perform speed-wise compared to the CLI, but I think that aiming for parity with the Gutenberg testing utils would be preferable in the long run. Since we're using wp-env as a testing env as well, we could even import the upstream [e2e testing utils](https://github.com/WordPress/gutenberg/tree/trunk/packages/e2e-test-utils-playwright) directly and use them OOTB.

I might, of course, be misunderstanding the intent here - I'm not even sure yet if we're already using the core e2e utils or not!

Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: > I might, of course, be misunderstanding the intent here - I'm not even sure yet if we're already using the core e2e utils or not!

@WunderBart you're good, I really appreciate your thoughts here and I agree after a minutes thought, matching with Gberg here makes sense and I can't imagine the perf being much different. I'll implement this. Thank you for this valuable suggestion. My biggest concern is authenticating the request because I don't know if these tests have prior art for that but i'm sure its solvable.

Thanks again its great feedback!

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts (hunk 9 above)
Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: > Because creating the post and tearing it down takes some time to do every single test. But I'm still keen to be convinced I'm wrong here, tbh I haven't done a lot of PlayWright testing so maybe I'm wrong.

I don't think you're wrong - it will likely take more time than just calling `page.goto`, but it's the kind of a tradeoff that usually needs to be considered: Do we want to wait a little longer but provide better resilience or is it ok to risk some flakiness for the sake of speed? There are many factors to it, but from my experience, e2e flakiness has never been welcomed by fellow devs! 😅 

Isolating tests has been a good practice stability-wise, e.g., in the core Gutenberg testing. To be fair, full isolation is still impossible since we're testing against a single wp-env instance, where the editor preferences are persistent (changing a preference will bleed into parallel tests). Ideally, though, each test should be completely isolated.

File: plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts (hunk 6 above)
Comment (dinhtungdu) [reply plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: > Maybe the rest API would perform better

@samueljseay  I believe it is. Using CLI we need some extra steps like spawning a node process and running the CLI command through `wp-env`. The later is painfully slow.

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts (hunk 9 above)
Comment (dinhtungdu) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: > Because creating the post and tearing it down takes some time to do every single test.

I think it won't be a big issue if we use REST API to manipulate posts.

> we could consider turning createPostFromTemplate into a [Playwright Fixture](https://playwright.dev/docs/test-fixtures)

💯  I will try to use Page Object Model until I can't 😄

File: plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts (hunk 6 above)
Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: > I'll implement this.

Fantastic! 🚀 

> My biggest concern is authenticating the request because I don't know if these tests have prior art for that but i'm sure its solvable.

Once we're logged in, the only thing that might be needed is the wp nonce attached to the request. I'd recommend checking against [Gutenberg's `rest.ts`](https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-test-utils-playwright/src/request-utils/rest.ts) fixture module implementation - should be enough to make this work! 🤞 

If there's anything more I can assist with, just let me know!

File: plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts (hunk 4 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/utils/create-dynamic-content.ts]: This is no longer relevant or needed.

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts (hunk 5 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: @dinhtungdu don't know what I was thinking, this doesn't make sense to do.

its because of how a file path is resolved in node.

if we pass relative path to readFile it will be determined by current working directory. cwd is based on where node was run from. 

if we pass a relative path from a test, we have no idea of the location of that test file. The test will want to resolve the filename from relative to the test file and not relative to cwd or anything else. You can pass a relative path if you want but it will be calculated from where node was run from and that could be different due to pnpm allowing you to run commands from anywhere in monorepo.

I can't right now think of a better alternative since I'd prefer to let templates go right next to tests rather than in a prescribed directory.

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts (hunk 9 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: Yep I think the REST API will make this fast enough not to matter. (perf was originally why i did this here in before / after.

**Comment by samueljseay** [comment]
@WunderBart thank you the fixtures and rest API made for great improvements, it happens that RequestUtils was already used so not an issue to utilize here. Would you mind taking another look?

🙇

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts (hunk 9 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: Resolved in https://github.com/woocommerce/woocommerce/pull/44223/commits/64b60b48aeacf4bc79d38f9133e1d05261e7344f

**Comment by WunderBart** [comment]
> it happens that RequestUtils was already used so not an issue to utilize here.

That's excellent, @samueljseay! It suggests new templating options for us to explore, assuming our primary aim is to align closely with upstream practices. Before reviewing the PR again, I might have a potential solution alternative to handlebars! 🙌 

#### **Creating block variant via the `Editor.insertBlock` API**.

The `insertBlock` function allows inputting a block's representation in JSON format, as defined in the `block.json` file. This approach provides a type-safe way to define block variants, bypassing the use of handlebars, the necessity to create new pages, or the need to call REST endpoints. A simple PoC for this would be as follows:

```js

function createStockStatusBlockRepresentation( { title, /* other attrs */ } ) {
  return {
    name: 'woocommerce/product-collection',
    attributes: { /* ... */ },
    innerBlocks: [
      {
        name: 'woocommerce/product-template',
        attributes: { /* ... */ },
        innerBlocks: [
          {
            name: 'core/post-title',
            attributes: { content: title },
          },
          // ...
        ],
      },
      // ...
    ]
  };
}

const defaultStockStatusBlockRepresentation = createStockStatusBlockRepresentation( { title: 'hello!' } );

test( 'Default Stock Status block', async ({ editor, page }) => {
  await editor.insertBlock( defaultStockStatusBlockRepresentation );

  // test as usual.
} );
```

You can see an example usage in Gutenberg [here](https://github.com/WordPress/gutenberg/blob/c7d0e6eaa80b732bddb294a3ed22ff66723a6fd4/test/e2e/specs/editor/blocks/columns.spec.js#L127-L143).

Do you think it makes sense? Do you see any limitations in this solution that handlebars would not have?

**Comment by samueljseay** [comment]
@WunderBart yeah we use that insertBlock API in some places where we're just testing a single block on its own, but, I think the limitation of that API is that its extremely cumbersome when you want a template of various blocks surrounded by the block under test.

We want templates so we can describe complex page and post layouts while also passing attributes to blocks under test. Sometimes the block may be first or sometimes it may be last nested in a bunch of other content.

That said, I am having a look at some old templates that certainly could have been replaced with `insertBlock`. It seems perhaps that we tend to copy/paste from each others examples when writing tests, leading to a mix of all kinds of approaches so it would also be good in future to consolidate and use `insertBlock` for single block under test.

**Comment by WunderBart** [comment]
> yeah we use that insertBlock API in some places where we're just testing a single block on its own, but, I think the limitation of that API is that its extremely cumbersome when you want a template of various blocks surrounded by the block under test.

Good point. Looking at the current stock status template, it would be difficult to manually create its representation for insertBlock, while for the handlebars template, it would just be a copypasta of the block code with added variables. 👍 

I'll continue with the review then! 👀

File: plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+		const outOfStockCheckbox = page.getByLabel( 'Out of stock' );
+		await outOfStockCheckbox.click();
+
+		// wait for navigation
+		await page.waitForURL( /.*filter_stock_status=outofstock.*/ );
+
+		const products = page.locator( '.wc-block-product' );
+
+		await expect( products ).toHaveCount( 1 );
+	} );
```
Comment (WunderBart) [inline plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: I think overriding the `post` fixture per the `describe` block (instead of defining new block fixtures per suite) could make the code more intuitive and easier to follow. For example:

```js
test.describe( 'Product Filter: Stock Status Block', async () => {
	test.describe( 'with default settings', () => {
		test.use( {
			page: async ( { page, requestUtils }, use ) => {
				const post = await createPostFromTemplate(
					requestUtils,
					{ title: 'Product Filter Stock Status Block' },
					TEMPLATE_PATH,
					{}
				);

				await page.goto( post.link );
				await use( page );
			},
		} );

		test( 'renders a checkbox list with the available stock statuses', async ( {
			page,
		} ) => {
			const stockStatuses = page.locator(
				'.wc-block-components-checkbox__label'
			);

			await expect( stockStatuses ).toHaveCount( 2 );
			await expect( stockStatuses.nth( 0 ) ).toHaveText( 'In stock' );
			await expect( stockStatuses.nth( 1 ) ).toHaveText( 'Out of stock' );
		} );

		test( 'filters the list of products by selecting a stock status', async ( {
			page,
		} ) => {
			const outOfStockCheckbox = page.getByLabel( 'Out of stock' );
			await outOfStockCheckbox.click();

			// wait for navigation
			await page.waitForURL( /.*filter_stock_status=outofstock.*/ );

			const products = page.locator( '.wc-block-product' );

			await expect( products ).toHaveCount( 1 );
		} );
	} );

	test.describe( 'with displayStyle set to dropdown', () => {
		test.use( {
			page: async ( { page }, use ) => {
			// ...
```

This way, each scoped test starts at the target `page`, and the variants separation is done at a parent level, which allows for a more concise test title.

You might also notice that I didn't call `deletePost` in this example. It's because I think it's better to just call `requestUtils.deleteAllPosts()` [in the global setup](https://github.com/woocommerce/woocommerce/blob/417683095a20177f45b6137881167fee162626f1/plugins/woocommerce-blocks/tests/e2e/global-setup.ts#L237-L244), so that it resets the env before any test is run. This way we can drop that extra `deletePost` implementation and likely save time on the execution.

Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: @WunderBart I thought the point is to keep tests isolated from each other? if we don't delete the post when the test runs its technically one more change spilling over between test suites? That said I'm fine not to delete them until restarting the suite as long as no one relies on the changed state of some existing post in another test.

There is a blocker for actually doing this though. The existing approach is that a set of posts is generated before global setup runs via wp-cli in bash scripts. (They happen as part of setting up the docker environment). So running `deleteAllPosts` at that point will delete the ones created during env setup. As you and I are discussing here, its pretty obvious why this approach is not good in its current form so I think we should move to the runtime generation of posts for all tests eventually and post,page,product creation for tests should happen during running of the suite instead, then we could adopt this approach of `deleteAllPosts`.

TLDR; Yes I agree your approach will be a great improvement but I think we have to do it as a follow up in future, refactoring away from this "at env start" approach to post generation towards the more dynamic approach we've created here. Thank you for the thoughtful feedback because imo this gives us a clear path of better practice / approach and we can add some more cooldown tasks to address these shortcomings in the current setup.

As for the scoping approach, it looks good to me! I agree its more intuitive to see the context of what's under test. I'll implement that! 🙇

Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: > I thought the point is to keep tests isolated from each other? if we don't delete the post when the test runs its technically one more change spilling over between test suites? 

Not sure I follow. With the fixture approach, each `test()` will receive a unique post the page will open with. Even if we don't delete them at all, they should not be bleeding over the other tests unless we're specifically testing, e.g., the number of posts or something in that area (which I don't think we're doing). In the initial approach we were running multiple tests against the same post, which was likely to produce unpredictable initial state. Anyway, in the suggested approach, I think we should be safe whether we remove the post or not! Does that make sense?

> There is a blocker for actually doing this though. The existing approach is that a set of posts is generated before global setup runs via wp-cli in bash scripts. (They happen as part of setting up the docker environment). So running deleteAllPosts at that point will delete the ones created during env setup. 

TIL! It does sound like a blocker. I think it's still worth pursuing though, as it's safer to cleanup in the global setup than in an after* hook, as they aren't always called. When cleaning in the after hook, all it takes to have a dangling post is to stop the test somewhere in the middle!

> As for the scoping approach, it looks good to me! I agree its more intuitive to see the context of what's under test. I'll implement that! 🙇

Nice, happy I could help! 🙌

Comment (WunderBart) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: As a potential follow-up, it would be great to implement the `createPostFromTemplate` as an extension of core's `RequestUtils` (possibly somewhere [here](https://github.com/woocommerce/woocommerce/blob/d329ced1e8aff7d5b3b4a63364930372519d1aa9/plugins/woocommerce-blocks/tests/e2e/playwright-utils/test.ts#L157-L167)). This way we wouldn't need to import it explicitly and the usage would be a tad simpler:

```js
test.use( {
	page: async ( { page, requestUtils }, use ) => {
		const post = await requestUtils.createPostFromTemplate(
			// no need to pass requestUtils as an argument
			{ title: 'Product Filter Stock Status Block' },
			TEMPLATE_PATH,
			{}
		);

		await page.goto( post.link );
		await use( page );
	},
} );
```

Comment (samueljseay) [reply plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts]: > In the initial approach we were running multiple tests against the same post, which was likely to produce unpredictable initial state. Anyway, in the suggested approach, I think we should be safe whether we remove the post or not! Does that make sense?

Yes for sure, I just assumed you were wanting more perfect isolation per test than that, yes in future someone could be testing something meta about posts or something else, but I am not concerned about it either. I just wanted to get on same page with you in this instance 👍🏻 

>  I think it's still worth pursuing though

Yes I'm advocating to do it, but it will require refactoring many other tests. I will create a cooldown task to do that refactoring.

> As a potential follow-up, it would be great to implement the createPostFromTemplate as an extension of core's RequestUtils (possibly somew[here](https://github.com/woocommerce/woocommerce/blob/d329ced1e8aff7d5b3b4a63364930372519d1aa9/plugins/woocommerce-blocks/tests/e2e/playwright-utils/test.ts#L157-L167) here). This way we wouldn't need to import it explicitly and the usage would be a tad simpler:

Great idea. I might not get a chance to work on this again for a couple of cycles so I'll do that in this PR

**Comment by samueljseay** [comment]
@WunderBart I don't think I want to make any changes here.

I tried the `test.use` approach and although it improves readability it required significant changes to delete the post. We can't delete the post in the `use` because then it won't be available for subsequent posts. I started writing something to keep track of the posts to delete them separately without affecting the ones created at env startup, but it just ended up feeling like a workaround for the fact that we need to address the issue we discussed of the templates that are created during setup. I don't think its worth complicating this solution right now to workaround the existing templates.

It's actually a cooldown period for me now, so if its ok with you, I'd like to merge this in it's current state and work on refactoring the templates that are created on startup. Let me know what you think.

**Comment by WunderBart** [comment]
> I tried the `test.use` approach and although it improves readability it required significant changes to delete the post. We can't delete the post in the `use` because then it won't be available for subsequent posts

I believe you meant "subsequent tests," which would be a good thing, actually! Each test should be unique, and if we need to make a multi-step test (against the same post), let's use [test.step](https://playwright.dev/docs/api/class-test#test-step) instead. Let me know if you meant something else 😅 

> It's actually a cooldown period for me now, so if its ok with you, I'd like to merge this in it's current state and work on refactoring the templates that are created on startup. Let me know what you think.

Absolutely! We can always address the aforementioned stuff in a follow-up PR if needed. 🚀 

_edit_

Since it's a cooldown period right now, I guess I could also try playing around with it, if that's OK with you! 🙌

**Comment by samueljseay** [comment]
@WunderBart 

> I believe you meant "subsequent tests," which would be a good thing, actually! Each test should be unique, and if we need to make a multi-step test (against the same post), let's use [test.step](https://playwright.dev/docs/api/class-test#test-step) instead. Let me know if you meant something else

Yes I meant subsequent tests, but I don't think its good. I'm grappling with exactly how these fixture overrides work, but, test.use used at the describe level, and overriding page doesn't work how you want it to here. it never recreates the post in any subsequent test. It runs once, deletes the post on the first test and subsequent tests do not seem to run this code against the page property. 

> Specifies options or fixtures to use in a single test file or a [test.describe()](https://playwright.dev/docs/api/class-test#test-describe-1) group. Most useful to set an option, for example set locale to configure context fixture.

This implies to me its run once and thats the behaviour I see. Do you know more about this? Also it feels like using this to set a fresh fixture every test is actually not the intended use of test.use.

**Comment by samueljseay** [comment]
> Since it's a cooldown period right now, I guess I could also try playing around with it, if that's OK with you! 🙌

I don't mind, but please let me know what you're working on, I intend to work on this effort too and just don't want to duplicate work 😄

**Review by WunderBart — APPROVED** [review]
LGTM! 🚀 

Thank you for addressing all the feedback! 🙇 Let's explore the other discussed solutions in a follow-up PR.  

[non-blocker]: We might want to [update the readme example](https://github.com/woocommerce/woocommerce/pull/44223/files#diff-38dbd051cc7466eb92caf1c34c888bdc8b628cec86cf898f39834ed4fc02b502R131-R175) to match the implementation.

**Comment by samueljseay** [comment]
Thanks @WunderBart I missed the readme. I'll update that 😄
